CSF: Allow overridding globals at the story level

tmeasday opened the pull request:
Closes: #23347 #27709 #27524 #27523 

## What I did

- Change the behavior of globals so that `userGlobals` and `storyGlobals` are separate in state.
  - `userGlobals` are globals set via toolbars/addons etc. These are dynamic and change-able.
  - `userGlobals` are initialized using `initialGlobals`.
  - `storyGlobals` are set via a `.globals` property on the `meta` or `story` inside `*.stories.*` files.
  - `storyGlobals` are fixed in place, and cannot be changed.
- `initialGlobals` is introduced as a concept.
  - These can only be set in a `preview.*` file. Addons can set them too.
  - `initialGlobals` are used to determine if a global's value should appear in the URL or not.
- `globals` is still available and results from shallow-merging `userGlobals` and `storyGlobals`.
- the `useGlobals` hook now return a third value: `storyGlobals`.
- decorators now receive an extra context property: `storyGlobals`.

---

- Adddon Viewport is configured slightly differently (see `MIGRATION.md`).
- Addon Backgrounds is configured slightly differently (see `MIGRATION.md`).
- Addon Themes now has an extra parameter `disable`, allowing the disabling of the tool per story.

## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:
- [x] stories
- [x] unit tests
- [ ] integration tests
- [x] end-to-end tests

#### Manual testing

1. Create & run a sandbox 
2. Visit `localhost:6006/?path=/story/lib-preview-api-globals--overrides` to see basics
3. Visit `localhost:6006/?path=/story/addons-toolbars-globals--override` to see toolbar behaviour
4. Test of the background & viewport addon (they use globals)
5. in the sandbox follow this migration: https://github.com/storybookjs/storybook/blob/6141b5b81a5693cf3eca7f926a6b513bf15b4cd3/MIGRATION.md#L429-L463
6. Now test backgrounds and viewports again
	- take note that the URL should not contain the property/value when the value matches the initialGlobal
	- you should be able to setup initialGlobals in `preview.ts` and change the default value. example:
	  ```ts
	  export const initialGlobals = { viewport: { value: "phone" } };
	  ```
7. Test the inheritance of globals by
	- setup a initialGlobal value
	- set the same global on a meta
	- set the same global on a story
	- expect to see the value set on the story
	- disable the value on the story, and expect the meta's value to be shown.
8. Test the new context values by creating a decorator:
	```ts
	export const decorators = [(StoryFn, { globals, storyGlobals, userGlobals }) => {
		console.log({ globals, storyGlobals, userGlobals });
		return Story();
	}];
    ```
	This should show the correct info when navigating stories, and changing globals


### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->
This pull request has been released as version `0.0.0-pr-26654-sha-49d01553`. Try it out in a new sandbox by running `npx storybook@0.0.0-pr-26654-sha-49d01553 sandbox` or in an existing project with `npx storybook@0.0.0-pr-26654-sha-49d01553 upgrade`.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-26654-sha-49d01553`](https://npmjs.com/package/storybook/v/0.0.0-pr-26654-sha-49d01553) |
| **Triggered by** | @ndelangen |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`tom/23347-story-globals`](https://github.com/storybookjs/storybook/tree/tom/23347-story-globals) |
| **Commit** | [`49d01553`](https://github.com/storybookjs/storybook/commit/49d015536481cc10ea380eb07aaeeaad9751f2fb) |
| **Datetime** | Fri Aug  2 11:46:36 UTC 2024 (`1722599196`) |
| **Workflow run** | [10215034516](https://github.com/storybookjs/storybook/actions/runs/10215034516) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=26654`_
</details>
<!-- CANARY_RELEASE_SECTION -->

tmeasday committed:
Add globalOverrides concept to CSF

tmeasday committed:
Override globals and emit old values on UPDATE_GLOBALS

tmeasday committed:
Always emit `UPDATE_GLOBALS` when selecting story

tmeasday committed:
Add `userGlobals` to manager globals code

tmeasday committed:
Disable toolbars when global is overridden

shilman committed:
Improve the test case to include autodocs with multiple overrides

shilman commented on code/addons/toolbars/src/components/ToolbarMenuList.tsx:
```diff
@@ -79,6 +82,7 @@ export const ToolbarMenuList: FC<ToolbarMenuListProps> = withKeyboardCycle(
       >
         <ToolbarMenuButton
           active={isTooltipVisible || hasGlobalValue}
+          disabled={isOverridden}
```
Readonly might be a better UX. See this recent PR

https://github.com/storybookjs/storybook/pull/26577

tmeasday replied:
Can we leave it to someone else (w/ @cdedreuille's help) to create a readonly version of the `TooltipLinkList` & `ToolbarMenuButton`? They doesn't appear to exist at the moment.

shilman reviewed (approved):
This is great! We might want to do the other addons in a separate PR or separate PRs to scrutinize each one a little more.

tmeasday committed:
Update code/addons/toolbars/src/components/ToolbarMenuList.tsx

Co-authored-by: Michael Shilman <shilman@users.noreply.github.com>

valentinpalkovic commented on code/addons/toolbars/src/components/ToolbarMenuList.tsx:
```diff
@@ -18,11 +19,12 @@ export const ToolbarMenuList: FC<ToolbarMenuListProps> = withKeyboardCycle(
 
     const currentValue = globals[id];
     const hasGlobalValue = !!currentValue;
+    const isOverridden = !deepEqual(currentValue, userGlobals[id]);
```
a deep equal check might be problematic for cases where functions or class instances are assigned to global values, leading to potential circular references.

Why can't we check whether `userGlobals[id]` is present? This matches my expectation of an "override". Otherwise, we should also rename the variable to something like `isDeepEqual`.

```suggestion
    const isOverridden = id in userGlobals;
```

tmeasday replied:
@valentinpalkovic:

> a deep equal check might be problematic for cases where functions or class instances are assigned to global values
globals need to be serializable to go over the channel, so I don't think your examples are possible. But there may be others we should consider.

> Why can't we check whether userGlobals[id] is present? 

So I think you have this backwards (this might be a sign that the naming is bad) - `userGlobals` are the globals the user has specified via clicking toolbars etc. `globals = { ...userGlobals, ...globalOverrides }`. So this is how we check if a `globalOverride` is set for a key (does `globals[key] == userGlobals[key]`).

Now I think about it, maybe we've made it difficult to do the thing that you generally would want to do (know if a value has been overrdden). Perhaps instead we should set `{ globals, globalOverrides }` on the event, or even `{ globals, globalOverrides, userGlobals }`?

WDYT @shilman

valentinpalkovic commented on code/lib/manager-api/src/index.tsx:
```diff
@@ -495,9 +496,9 @@ export function useArgs(): [Args, (newArgs: Args) => void, (argNames?: string[])
 }
 
-export function useGlobals(): [Args, (newGlobals: Args) => void] {
+export function useGlobals(): [Globals, (newGlobals: Globals) => void, Globals] {
```
```suggestion
export function useGlobals(): [
  globals: Globals,
  updateGlobals: (newGlobals: Globals) => void,
  userGlobals: Globals,
] {
```

tmeasday committed:
Update code/lib/manager-api/src/index.tsx

Co-authored-by: Valentin Palkovic <valentin@chromatic.com>

kylegach reviewed (approved):
Fwiw, I'm good with the `globalOverrides` name. It's perhaps a bit clunky, relative to `globals`, but I think it's necessary to communicate that the behavior is different.

@tmeasday — Should this PR contain docs?

tmeasday committed:
Upgrade CSF to version w/ globals on context

From https://github.com/ComponentDriven/csf/pull/95

tmeasday committed:
Update API to include `storyGlobals`

This means we don't need to do any complex checks in addon toolbars to tell if it should be disabled

tmeasday committed:
Also emit storyGlobals on update

tmeasday committed:
Updated existing tests

tmeasday committed:
Update integration tests to account for various storyGlobals scenarios

tmeasday committed:
Add end-to-end tests of toolbars/overrides

tmeasday committed:
Fix unused random import

valentinpalkovic commented:
What are the implications on the docs page? Currently, it seems that globally set globals via the toolbar either don’t have an effect, are invisible in doc mode, or are applied to all stories simultaneously. Is it considered to apply story-specific globals also on the docs page (especially for autodocs)? How do we/addon authors communicate if particular globals only apply in the story view and not in the docs view?

tmeasday commented:
@valentinpalkovic -- there's two angles on that:

1. Do/which globals apply in each story in docs mode? The answer is that globals are defined once for the whole SB, so the same values would apply for each story.

2. If a story sets `story.globals`, what happens in docs mode? 
i.	Currently `story.globals` only triggers a `GLOBALS_UPDATED` event (and thus changes a toolbar for instance) if the story is **selected** (ie in story mode).
ii. However, the story itself will get rendered (in docs mode) with the overridden global. You can see that if you visit http://localhost:6006/?path=/docs/lib-preview-api-globals--docs in this PR.

tmeasday committed:
Update `globals` test

tmeasday committed:
Fix url test

tmeasday committed:
Update snapshots in StoryStore tes

tmeasday committed:
Update snapshots in normalize story test

valentinpalkovic commented on code/lib/preview-api/src/modules/preview-web/PreviewWithSelection.tsx:
```diff
@@ -399,6 +410,9 @@ export class PreviewWithSelection<TRenderer extends Renderer> extends Preview<TR
       });
+      // We need to update globals whenever we go in or out of an overridden story.
+      // As an optimization we could check if that's the case, but it seems complex and error-prone
+      this.channel.emit(GLOBALS_UPDATED, { userGlobals, storyGlobals, globals, initialGlobals });
```
Also emit "GLOBALS_UPDATED" when visiting docs page (else block)

tmeasday commented on code/addons/toolbars/src/components/ToolbarMenuList.tsx:
```diff
@@ -18,11 +18,12 @@ export const ToolbarMenuList: FC<ToolbarMenuListProps> = withKeyboardCycle(
     toolbar: { icon: _icon, items, title: _title, preventDynamicIcon, dynamicTitle },
   }) => {
-    const [globals, updateGlobals] = useGlobals();
+    const [globals, updateGlobals, storyGlobals] = useGlobals();
```
This is the API that addons use to tell if the user has overridden a global value they are interested in.

If `storyGlobals[X]` is set, then the value of `globals.X` is hard-coded (for now) and so showing a read/write UI is probably not sensible.

ndelangen committed:
Merge branch 'next' into tom/23347-story-globals

ndelangen committed:
use canary of csf

ndelangen committed:
add missing `storyGlobals` to test object

ndelangen committed:
add core templates to ui storybook

ndelangen commented:
Might there be a bug somewhere that's causing this to happen:
> Attempted to set a global (foo) that is not defined in initial globals or globalTypes

I've added the core template stories to the ui storybook, because that seemed useful.
The globals tests stories work great in the sandbox, but not in the ui storybook.
And the difference is the warning in the browser console above.

ndelangen committed:
remap core template stories a little

ndelangen committed:
add extra initialGlobals & globalTypes

ndelangen commented on code/.storybook/preview.tsx:
```diff
@@ -312,4 +314,11 @@ export const globalTypes = {
       ],
     },
   },
+  foo: { defaultValue: 'fooDefaultValue' },
+  bar: { defaultValue: 'barDefaultValue' },
+};
+
+export const initialGlobals = {
+  foo: 'fooValue',
+  baz: 'bazValue',
 };
```
@tmeasday Can I get an explainer why `defaultValue: 'fooDefaultValue'` should ever be used?
How are initialGlobals.foo and globalTypes.foo.defaultValue connected?

At first glance, me and @valentinpalkovic both concluded that if you set a `initialGlobal.foo`, then `globalTypes.foo.defaultValue` should never be used?

ndelangen replied:
I was able to figure it out!

ndelangen committed:
fix mapping

ndelangen committed:
fix

tmeasday commented:
@ndelangen 

> Might there be a bug somewhere that's causing this to happen:

 >> Attempted to set a global (foo) that is not defined in initial globals or globalTypes

It sounds like you added a story that sets `globals.foo` but there is no global of that name defined in `preview.ts` or an addon. This is a restriction we added to globals for reasons I don't really recall :)

ndelangen committed:
use the feature ourselves for theme

ndelangen committed:
remove defaultValue

ndelangen committed:
cleanup

kasperpeulen commented on code/core/src/preview-api/modules/store/csf/prepareStory.ts:
```diff
@@ -227,6 +233,7 @@ function preparePartialAnnotations<TRenderer extends Renderer>(
     parameters,
     initialArgs: passedArgs,
     argTypes: passedArgTypes,
+    storyGlobals,
```
I would just call this `globals`.

kasperpeulen replied:
> The divergence in types between PreparedStory and StoryContext definitely has the potential to be confusing.

IMO, they should diverge. I think with CSF4 we should call them just `Story` and `StoryContext` and `Story` is what we return from the csf factories:

```tsx
const Basic = defineStory({ // <-- name up to debate
  args: { ... }
})
```

And `StoryContext` is what we provide as argument to play, render, decorators etc. and it has 
* extra stuff like `canvasElement` and `abortSignal`
* will resolve the `context` argument for "properties/fixtures" like `mount`, `renderToCanvas`, `step`
* integration such as storybook can override values such as `args` and `globals` in their UI

> Maybe we shouldn't make PreparedStory extend StoryContext?
Probably PreparedStory should extend StoryAnnotations and ProjectAnnotations? It is kind of the resolved version of that.

In some sense, the relationship is kind of the opposite at the moment at runtime, StoryContext extends PrepareStory. But because of that, this stuff:
https://github.com/storybookjs/storybook/blob/c11a299905f3fd3e77f63552e2bcde2b2180e4fe/code/lib/types/src/modules/story.ts#L100-L108

Actually sneakily ends up on the story context on runtime
![image](https://github.com/user-attachments/assets/04bda0d1-4263-4c44-b88b-f40f42ed31f7)

We should probably get rid of that?

tmeasday replied:
Agree with all of that.

ndelangen replied:
@kasperpeulen what's the action that follows out of this discussion that needs to happen to this PR?

kasperpeulen commented on code/core/src/preview-api/modules/store/StoryStore.ts:
```diff
@@ -230,10 +230,17 @@ export class StoryStore<TRenderer extends Renderer> {
+      userGlobals,
+      globals: {
+        ...userGlobals,
+        ...story.storyGlobals,
```
so this could be `story.globals`

ndelangen replied:
Can you explain?

ndelangen committed:
enhancements fro background & viewports

ndelangen committed:
fixes for viewports & backgrounds

ndelangen committed:
Merge branch 'next' into tom/23347-story-globals

ndelangen committed:
duplicate rather than complicate

ndelangen committed:
re-use the previewAnnotation files

ndelangen committed:
fix e2e test

ndelangen committed:
fixes

ndelangen committed:
cleanup

ndelangen committed:
disable the problematic stories for now

ndelangen committed:
fix https://www.chromatic.com/test?appId=635781f3500dd2c49e189caf&id=66993117d68df8aba39b0aed

ndelangen committed:
oops

ndelangen committed:
add comment for future me

ndelangen committed:
cleanup

ndelangen committed:
add test exclusion tag

ndelangen committed:
cleanup

ndelangen committed:
Merge branch 'next' into tom/23347-story-globals

ndelangen committed:
move `storyGlobals` type to types

as discussed here:
https://github.com/ComponentDriven/csf/pull/95#issuecomment-2234809100

ndelangen committed:
remove duplicate

ndelangen committed:
fix the `globe` text showing in the reset of a toolbar

ndelangen committed:
add icons ref

ndelangen committed:
restructure the components

ndelangen committed:
restructure the main storybook a bit

ndelangen committed:
fix

ndelangen committed:
add clear deprecation status to deprecated icons

ndelangen committed:
fix

ndelangen commented on code/addons/toolbars/src/components/ToolbarMenuList.tsx:
```diff
@@ -18,11 +18,12 @@ export const ToolbarMenuList: FC<ToolbarMenuListProps> = withKeyboardCycle(
 
     const currentValue = globals[id];
     const hasGlobalValue = !!currentValue;
+    const isOverridden = id in storyGlobals;
```
What's the conclusion on this topic?

ndelangen committed:
Merge branch 'next' into tom/23347-story-globals

ndelangen committed:
fix the switching to docs not resetting storyGlobals

ndelangen committed:
add a version of viewport tool that uses globals exclusively

ndelangen committed:
improve the viewports addon

ndelangen committed:
wip

ndelangen committed:
cleanup

ndelangen committed:
cleanup

ndelangen committed:
restructuring

ndelangen committed:
add

ndelangen committed:
Merge branch 'next' into tom/23347-story-globals

ndelangen committed:
change back to use parameters for addon config

ndelangen committed:
Merge branch 'next' into tom/23347-story-globals

ndelangen committed:
fixes

ndelangen committed:
add more

ndelangen committed:
fix

ndelangen committed:
fixing typings

ndelangen committed:
Merge branch 'next' into tom/23347-story-globals

ndelangen committed:
always ensure the storyStore is ready to updateGlobals before proceeding to update globals

ndelangen committed:
cleanup

ndelangen committed:
add disabled: false to default parameters of backgrounds

ndelangen committed:
fix link to moved story

ndelangen committed:
add work-around hopefully reducing e2e test flake

ndelangen committed:
testing a fix with yann

Co-authored-by: Yann Braga <yannbf@gmail.com>

ndelangen committed:
Merge branch 'next' into tom/23347-story-globals

ndelangen committed:
add retry for playwright test that is failing

simplify test

Co-authored-by: Yann Braga <yannbf@gmail.com>

ndelangen committed:
Merge branch 'tom/23347-story-globals' of https://github.com/storybookjs/storybook into tom/23347-story-globals

ndelangen committed:
disable webkit in e2e tests for now,due to flake:

https://app.circleci.com/pipelines/github/storybookjs/storybook/81120/workflows/16a6a303-ed49-45c8-be74-769ff0cf1253/jobs/686641/artifacts

ndelangen committed:
Merge branch 'next' into tom/23347-story-globals

ndelangen committed:
ignore CPC moving code in blame history

ndelangen committed:
simplify addon-toolbars, enable configuring via parameters, add feature-flag

ndelangen committed:
Merge branch 'next' into tom/23347-story-globals

shilman commented on code/addons/backgrounds/src/types.ts:
```diff
@@ -14,20 +21,33 @@ export interface BackgroundSelectorItem {
+
+export interface Config {
+  options: BackgroundMap;
+  disabled: boolean;
```
Can we please call this `disable` so that it's consistent with everything else?

shilman commented on code/addons/viewport/src/components/Tool.tsx:
```diff
@@ -0,0 +1,190 @@
+          title="Change the size of the preview"
+          active={isActive}
+          onDoubleClick={() => {
+            updateGlobals({ [KEY]: { value: undefined, isRotated: false } });
```
Do we have types for this?

ndelangen committed:
fix URL state for globals

ndelangen committed:
Revert "simplify addon-toolbars, enable configuring via parameters, add feature-flag"

This reverts commit 86db1e6252a5dd747dbd498126ad80e57a175e9e.

ndelangen committed:
Merge branch 'next' into tom/23347-story-globals

ndelangen committed:
make the API change to viewport's globals/parameters addon opt-in via feature flag

ndelangen committed:
add extra story testing disabled in new pattern

ndelangen committed:
clarify the naming/color

ndelangen committed:
add initialGlobals for toolbars, simplify decorator in main storybook

ndelangen committed:
override theme in chromatic or when playFunction, fixing some padding issues

ndelangen committed:
add more stories to toolbars stories

ndelangen committed:
improve typing

ndelangen committed:
consistent naming of disable, stronger typing on setting state

ndelangen committed:
cleanup

ndelangen committed:
cleanup

ndelangen committed:
fix test, after story rename/addition

ndelangen committed:
fix

ndelangen committed:
set initialGlobals of background to null

ndelangen committed:
chromatic preferred theme is stacked, not side-by-side

ndelangen committed:
rename the theme global we use internally to `sb_theme`. add addon-themes to template stories, and to main storybook.

ndelangen committed:
fix lockfile

ndelangen committed:
fix

ndelangen committed:
improvements

ndelangen committed:
improve

ndelangen committed:
improve

ndelangen committed:
improve

ndelangen committed:
improve

ndelangen committed:
improvement

ndelangen committed:
improvement

ndelangen committed:
improvements

ndelangen committed:
update docs

valentinpalkovic committed:
Cleanup

valentinpalkovic committed:
Cleanup

ndelangen committed:
Merge pull request #28765 from storybookjs/norbert/addon-themes-globals

CSF: Add addon-themes for testing to sandboxes & main storybook

ndelangen committed:
Merge branch 'next' into tom/23347-story-globals

ndelangen committed:
Merge branch 'tom/23347-story-globals' of https://github.com/storybookjs/storybook into tom/23347-story-globals

ndelangen committed:
Merge branch 'next' into tom/23347-story-globals
